interactions/api/gateway/client.py

```diff
@@ -112,160 +132,182 @@ def __init__(
+        """
+        return self._last_ack - self._last_send
+
+    async def run_heartbeat(self) -> None:
+        """Controls the heartbeat manager. Do note that this shouldn't be executed by outside processes."""
+
+        if self.__heartbeat_event.is_set():  # resets task of heartbeat event mgr loop
+            # Because we're hardresetting the process every instance its called, also helps with recursion
+            self.__heartbeat_event.clear()
+
+        if not self.__heartbeater.event.is_set():  # resets task of heartbeat ack event
+            self.__heartbeater.event.set()
+
+        try:
+            await self._manage_heartbeat()
+        except Exception:
+            self._closing_lock.set()
+            log.exception("Heartbeater exception.")
 
     async def _manage_heartbeat(self) -> None:
         """Manages the heartbeat loop."""
-        while True:
-            if self._closed:
-                await self.__restart()
-            if self.__heartbeater.event.is_set():
-                await self.__heartbeat()
-                self.__heartbeater.event.clear()
-                await sleep(self.__heartbeater.delay / 1000)
-            else:
+        log.debug(f"Sending heartbeat every {self.__heartbeater.delay / 1000} seconds...")
+        while not self.__heartbeat_event.is_set():
+
+            log.debug("Sending heartbeat...")
+            if not self.__heartbeater.event.is_set():
                 log.debug("HEARTBEAT_ACK missing, reconnecting...")
-                await self.__restart()
-                break
-
-    async def __restart(self) -> None:
-        """Restart the client's connection and heartbeat with the Gateway."""
-        if self.__task:
-            self.__task.cancel()
-        self._client = None  # clear pending waits
-        self.__heartbeater.event.clear()
-        await self._establish_connection(self.__shard, self.__presence)
-
-    async def _establish_connection(
-        self,
-        shard: Optional[List[Tuple[int]]] = MISSING,
-        presence: Optional[ClientPresence] = MISSING,
-    ) -> None:
-        """
-        Establishes a client connection with the Gateway.
+                await self._reconnect(True)  # resume here.
 
-        :param shard?: The shards to establish a connection with. Defaults to ``None``.
-        :type shard?: Optional[List[Tuple[int]]]
-        :param presence: The presence to carry with. Defaults to ``None``.
-        :type presence: Optional[ClientPresence]
+            self.__heartbeater.event.clear()
+            await self.__heartbeat()
+
+            try:
+                # wait for next iteration, accounting for latency
+                await wait_for(
+                    self.__heartbeat_event.wait(), timeout=self.__heartbeater.delay / 1000
+                )
+            except TimeoutError:
+                continue  # Then we can check heartbeat ack this way and then like it autorestarts.
+            else:
```

Comment:
What this looks like is the loop only continues if the `wait_for` times out, otherwise if the event gets set, then break the loop. Is that right?

Reply:
yeah pretty much, if it times out it relies on the unchanged event variables to correct itself via reconnection/etc for signals

Reply:
Pr will merge after resolving this @DeltaXWizard @Catalyst4222